### PR TITLE
Cache ccache across CI workflows

### PR DESCRIPTION
**Pull request by krcroft**
This PR uses ccache in the CI workflows, and more importantly caches its results.

The cache is saved uniquely for each job at the start of each day. Subsequent jobs of the same type (performed later in the day) will benefit from this first cache; and the only remaining work will be compiling the balance (or delta) of changed files. **Edit**: further improvement to this described in follow up comments below.

Here are the typical speed-ups after the first job of the day writes the cache:

| Workflow  | Job  | Before  | After  | Comments  |
|:--:|:--:|:---:|:---:|:---|
| Analysis  | Linters  | 50s  | n/a  | Cannot cache because doesn't involve compiling |
| Analysis  | Clang static  | [363s](https://github.com/dreamer/dosbox-staging/runs/526186138?check_suite_focus=true)  | [230s](https://github.com/dreamer/dosbox-staging/runs/526155560?check_suite_focus=true)  |   |
| Analysis  | PVS static  | [358s](https://github.com/dreamer/dosbox-staging/runs/526186157?check_suite_focus=true)  | n/a  | Cannot cache because analysis requires `strace`'ing GCC during compilation, which ccache circuments  |
| Analysis  | Clang sanitizers  | [121s](https://github.com/dreamer/dosbox-staging/runs/526186143?check_suite_focus=true)  | [71s](https://github.com/dreamer/dosbox-staging/runs/526155551?check_suite_focus=true)  |   |
| Analysis  | GCC sanitizers  | [250s](https://github.com/dreamer/dosbox-staging/runs/526186147?check_suite_focus=true)  |  [82s](https://github.com/dreamer/dosbox-staging/runs/526155555?check_suite_focus=true) |   |
| Linux  | GCC-5  | [81s](https://github.com/dreamer/dosbox-staging/runs/526183152?check_suite_focus=true)  | [31s](https://github.com/dreamer/dosbox-staging/runs/526154575?check_suite_focus=true)  |   |
| Linux  | GCC-7  | [94s](https://github.com/dreamer/dosbox-staging/runs/526183160?check_suite_focus=true)  | [67s](https://github.com/dreamer/dosbox-staging/runs/526154577?check_suite_focus=true)  |   |
| Linux  | GCC-9  | [112s](https://github.com/dreamer/dosbox-staging/runs/526183174?check_suite_focus=true)  | [70s](https://github.com/dreamer/dosbox-staging/runs/526154579?check_suite_focus=true)  |   |
| Linux  | Clang-8  | [98s](https://github.com/dreamer/dosbox-staging/runs/526183182?check_suite_focus=true)  | [63s](https://github.com/dreamer/dosbox-staging/runs/526154583?check_suite_focus=true)  |   |
| Linux  | Release  | [149s](https://github.com/dreamer/dosbox-staging/runs/526183214?check_suite_focus=true)  | [54s](https://github.com/dreamer/dosbox-staging/runs/526154595?check_suite_focus=true)  |   |
| macOS  | Clang  | [104s](https://github.com/dreamer/dosbox-staging/runs/526183168?check_suite_focus=true)  | [84s](https://github.com/dreamer/dosbox-staging/runs/526154624?check_suite_focus=true)  |   |
| macOS  | GCC-9  | [124s](https://github.com/dreamer/dosbox-staging/runs/526183176?check_suite_focus=true)  |  [90s](https://github.com/dreamer/dosbox-staging/runs/526154630?check_suite_focus=true) |   |
| macOS  | Release  | [224s](https://github.com/dreamer/dosbox-staging/runs/526183144?check_suite_focus=true)  | [183s](https://github.com/dreamer/dosbox-staging/runs/526154615?check_suite_focus=true)  |   |
| Windows  | Clang 32  | [652s](https://github.com/dreamer/dosbox-staging/runs/526154606?check_suite_focus=true)  | [178s](https://github.com/dreamer/dosbox-staging/runs/526211149?check_suite_focus=true)  |   |
| Windows | Clang 64  | [661s](https://github.com/dreamer/dosbox-staging/runs/526154607?check_suite_focus=true)  | [211s](https://github.com/dreamer/dosbox-staging/runs/526211151?check_suite_focus=true)  |   |
| Windows |  GCC 32  |  [680s](https://github.com/dreamer/dosbox-staging/runs/526154611?check_suite_focus=true) | [166s](https://github.com/dreamer/dosbox-staging/runs/526211153?check_suite_focus=true)  |   |
| Windows  | GCC 64  |  [678s](https://github.com/dreamer/dosbox-staging/runs/526154613?check_suite_focus=true) | [191s](https://github.com/dreamer/dosbox-staging/runs/526211161?check_suite_focus=true)  |   |
| Windows  | MSVC Debug  | [303s](https://github.com/dreamer/dosbox-staging/runs/526154619?check_suite_focus=true)  | n/a  | Cannot cache because not compatible w/ MSVC  |
|  Windows | MSVC Release  | [358s](https://github.com/dreamer/dosbox-staging/runs/526154638?check_suite_focus=true)  | n/a | Cannot cache because not compatible w/ MSVC  |

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -33,16 +48,36 @@ jobs:
   build_macos_release:
     name: Release build
     runs-on: macos-latest
+    env:
+      CC: ccache clang
+      CXX: ccache clang++
+      FLAGS: -O3 -DNDEBUG -pipe -march=nehalem
```
Comment (krcroft) [inline .github/workflows/macos.yml]: Moving these here means we use ccache and the optimization flags to build Opus as well :-)

**Comment by krcroft** [comment]
Some more small improvements:

- Now uses a two-day rolling cache where we fallback to yesterday's cache if we haven't yet generated one today, so the first build of the day doesn't start cold. 

- The catch here is that the cache can grow unabated because we're now bringing in yesterday's cache to today's results.  So we cap each cache to 64MB; which is more than enough (a single full build only uses about 15MB).

- The rolling cache also lets us keep our cache statistics, as opposed to being reset every day, so I've included a short summary at the end of the build script (if ccache is being used):



- Re-tightened up white-space and Log-env ordering that I introduced in prior commits.

(feel free to squash this PR down to one)

File: .github/workflows/analysis.yml [hunk 2]
```diff
@@ -38,6 +42,22 @@ jobs:
...
         run:  sudo pip3 install scan-build beautifulsoup4 html5lib
+
+      - name:  Prepare compiler cache
+        id:    prep-ccache
+        shell: bash
+        run:   |
+          mkdir -p "${CCACHE_DIR}"
+          echo "::set-output name=dir::$CCACHE_DIR"
+          echo "::set-output name=today::$(( $(date +%s) / 86400 ))"
+          echo "::set-output name=yesterday::$(( $(date +%s) / 86400 - 1 ))"
```
Comment (dreamer) [inline .github/workflows/analysis.yml]: I was looking at `date` options looking for alternatives and came up with some ideas like using `date +%y%j`, but your solutions will work better (no cache miss when new year rolls over).

I discovered there's an option of `date --date=yesterday +%Y%M%d`, but I don't know if it works cross-platform or only with new GNU - maybe worth checking out?

Comment (krcroft) [reply .github/workflows/analysis.yml]: Oh, nice find with the --yesterday flag! Will give it a try.

File: .github/workflows/windows.yml [hunk 3]
```diff
@@ -55,6 +56,21 @@ jobs:
...
         shell: python scripts\msys-bash.py {0}
         run:   ./.github/scripts/shrink-msys2.sh
+      - name:  Prepare compiler cache
+        id:    prep-ccache
+        shell: bash
+        run:   |
+          mkdir -p "${CCACHE_DIR}"
+          echo "::set-output name=dir::$CCACHE_DIR"
+          echo "::set-output name=today::$(( $(date +%s) / 86400 ))"
+          echo "::set-output name=yeserday::$(( $(date +%s) / 86400 - 1 ))"
```
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/windows.yml]: Oh; I thought I mass-fixed with sed in earlier commit. Will double-check and fix - thanks for pointing it out!

**Comment by krcroft** [comment]
> @krcroft can we merge #237 first, and then rebase this PR on top of master? I want to test if the solution for duplicated jobs works correctly for events, that were not triggered by me.

Go for it :rocket:

**Comment by dreamer** [comment]
OK, merged it in; now it's time for rebase - let's see if it will result in 19 checks or 38.

**Comment by dreamer** [comment]
Seems like everything is working nicely. Yay to the faster CI! Thanks! :)

**Comment by krcroft** [comment]
And no duplicate CI jobs for 3rd party PRs;  less work all-round!

**Pull request merged**